### ISSUE TITLE
Make BuildAbstractSyntaxTree method public

### ISSUE THREAD

**Issue by FabianNitsche** [adletec/sonic#46]
First of all: Thank you for taking over and pulling my original bug fixes in here and for fixing my bug fix... :D And even improving on the original implementation.

This is a feature request:
I would like to investigate my expression. The best way is to use the AbstractSyntaxTree. Please allow BuildAbstractSyntaxTree to be public at Evaluator.

I will provide a PR for your consideration.

**Comment by vlow** [comment]
Hi Fabian, Thank you for finding and fixing the bugs in the first place :)

Regarding your request: I'd like to get a better understanding of what you'd like to use the AbstractSyntaxTree for, before proposing a solution.

At the moment, I'm not convinced that we should expose the AST as part of the public API, for couple of reasons. This might sound inconsiderate when taking into account that the functionality is already there and you're adding the desired funtionality in a rather small PR. But my objection is stemming from architectural considerations, so please allow me to be a bit more verbose in my answer:

Before going more into detail, when I'm talking about public API, I'm refering to namespaces rather than visibility modifiers. So in an architectural sense, every type residing in the top level namespace (`Adletec.Sonic`) is part of the public API, e.g.  `Evaluator`. Types residing in sub packages (e.g. `AstBuilder` and `Operation`) are not.

For me, the main considerations behind what's part of public API and what's internal code is 

a) whether the type provides functionality or information which is part of the designated feature set and valuable on its own
b) whether we might (repeatedly) need to change the API in the future or keep the API stable

So if a type provides some valuable functionality for the typical _sonic_ user and we're confident that we can keep its API stable, we can make it part of the public API. By doing so, we must also provide types which are comprehensible to the user without detailed knowledge of the implementation.

Now when looking at your desired change to expose the AST, it's not sufficient to make `BuildAbstractSyntaxTree` public (in the _visibility modifier` sense). We would also need to expose the `Operation` class (in the `architectural` sense) and all of its subtypes, since this is what `BuildAbstractSyntaxTree` returns.

While I'm not entirely convinced, I might follow the notion that while exposing the AST is certainly not a valuable addition for the typical _sonic_ user, the functionality is already there and it will be valuable to some users (e.g. you 😀).

What's more important is "consideration b)": `Operation` and types derived from it are clearly part of the implementation and we need to retain the flexibility to change them whenever we see fit. But without creating a dedicated type for the AST which we can stable as part of the public API, we might break someones code with every change we make to the internal AST we do in the future.

**tl;dr**: adding the AST to the public API of _sonic_ will require to add a dedicated and stable data structure. Since _sonic_'s core promise is rapid evaluation rather than being a handy tool for AST parsing, I'm afraid this would be a lot of work for little benefit when done correctly.

So what is it you're using the AST for? Can you tell me more about your use-case? Maybe we can find an easier solution?

**Comment by FabianNitsche** [comment]
Thank you for your answer, I do understand your hesitation to add the `BuildAbstractSyntaxTree` to the public API. So let me give you a more complete picture of why and also my opinion on the architecture and what is already part of it. In the end it is your call to make.

Use case: I want to analyse a formula to substitute scalars for parameters to be able to fit them. So basically, I start with something like this:
`1.5 + 2 * x + 5 * x^2`
And I want to identify all that is not `x` and replace with parameters
`a0 +  a1 * x + a2 * x^a3`
Now I can (partially) fit them to data.
The AST is a powerful tool to manipulate formulas.

Now more to the way:
Yes I do not have to add it to `IEvaluator`. I could just use the classes from the infrastructure to do the job, but then I have two separtely configured toolchains. The one is the Parser, where there may be different functions or constants defined. And maybe there are also other different settings. I would have to keep them in sync. When I can use the same `IEvaluator` implementation, I know that the same thing happens when I parse a fromula to evaluate or when I build the AST. So they are consistent.

Now for the public API:
(I know the following under debate)
Everything that is not private or internal is public. One could even argue, that only private is not public, since C# allows some voodoo to be friendly with an assembly and use the internal stuff. I personally also dislike APIs that try to hide things and rather prevent extension - that said, I am aware that extensibility does come at a high cost, you already addressed. But do not think, that only by putting stuff into subfolders means that this is not part of the public API. It just structures your public API, stating: The top level stuff is for all of you, when you know what you are doing, look into the sublevels.

I think there are better ways of hiding implementation to create a stable API while allowing for change, than to reduce the API to the bare minimum. This makes usability and extensibility hard. Users may resort to Reflection to do what they want to do. And then basically everything is public... :D

Please do not get me wrong, I do not want to push you in my direction. I do understand your concerns. There may be a better way for allowing to take advantage of the AST as a user, maybe in a different way. My PR was just a suggestion and you should take it as that. I am perfectly fine with you dismissing this request. (But I may resort to Reflection... :D)

**Comment by vlow** [comment]
> I personally also dislike APIs that try to hide things and rather prevent extension [...]
> Everything that is not private or internal is public. One could even argue, that only private is not public, since C# allows some voodoo to be friendly with an assembly and use the internal stuff. [...]
> But do not think, that only by putting stuff into subfolders means that this is not part of the public API. It just structures your public API, stating: The top level stuff is for all of you, when you know what you are doing, look into the sublevels.

Agree. My intention isn't to keep anyone from using the 'internals'. It's to make a clear distinction between what is considered "safe to use for all of you" and "this is the inner workings. don't complain if it breaks between releases". That's why it's internal in the sense of structure, not visibility. Depending on how drastic changes are, I might still feel obligated to deprecate signatures before breaking them, but considering the Nuget download numbers and 11 GitHub stars, I feel pretty confident that I don't have to overthink my breaking changes yet 😁

> I think there are better ways of hiding implementation to create a stable API while allowing for change, than to reduce the API to the bare minimum. This makes usability and extensibility hard. Users may resort to Reflection to do what they want to do.

Again, agree. That's not my intentation. I'm not considering the AST part of what the typical user of the Evaluator is looking for, so it's not top level. It's still useful to some, so its visibility is `public`. Same for `Operation`. You'll find that `private` and `internal` modifiers are (apart from errors or technical debt) only used in locations which are only meaningful in the context of their (internal) usage, or if they'd expose data types of a different module.

> I could just use the classes from the infrastructure to do the job, but then I have two separtely configured toolchains. The one is the Parser, where there may be different functions or constants defined. And maybe there are also other different settings. I would have to keep them in sync.

Understood, that's sensible. The AstBuilder doesn't have many dependencies - just needs read access to function and constant definitions. Both are actually already part of the `IEvaluator` (eventhough I just discovered that `FunctionInfo` / `ConstantInfo` are in the `Execution` namespace, which makes me sound pretty hypocritical come to think about it 😅).

So if your focus is on the AstBuilder, we could easily expose it's dependencies without making `Operation` top level API or major changes to the `Evaluator`. Would you prefer this to reflection :D ?

**Comment by FabianNitsche** [comment]
> So if your focus is on the AstBuilder, we could easily expose it's dependencies without making Operation top level API or major changes to the Evaluator. Would you prefer this to reflection :D ?

How would you do that?

I do not find exposing some of the internal stuff as problematic. Actually, having FunctionInfo and ConstanInfo, enables me to at least to build up a `FunctionRepository` and `ConstantRepository` that is consistent when building the AST. I do not really agree, that this is exposing too much of the inner workings, because for people that do not care, this does not hurt, and for people that do need it, it makes it more usefu.. Do you intend to ever change that way of behavior? I think you may need to commit to some general parts of the library. If you ever think: "Meh, AST was the wrong horse", you probably have to create a different library anyway.

**Comment by FabianNitsche** [comment]
> considering the Nuget download numbers and 11 GitHub stars

This may not follow a linear equation... This can change pretty fast... :D
Peter and you did a great job.

**Comment by vlow** [comment]
> How would you do that?

Adding interfaces for IFunctionRegistry and IConstantRegistry which only contain read functionality and make the AstBuilder rely on them rather than the full r/w interfaces.

Example: 

```csharp
    public interface IConstantLookupProvider : IEnumerable<ConstantInfo>
    {
        ConstantInfo GetConstantInfo(string constantName);
        bool IsConstantName(string constantName);
    }
```

`Evaluator:Constants` would then return this interface instead of the plain IEnumerable, and AstBuilder would take it as constructor argument instead of IConstantRegistry. So you can get them in sync easily like that:

```csharp
var engine = Evaluator.CreateWithDefaults();
var astBuilder = new AstBuilder(engine.Functions, engine.Constants);
```

> Do you intend to ever change that way of behavior? I think you may need to commit to some general parts of the library. If you ever think: "Meh, AST was the wrong horse", you probably have to create a different library anyway.

I think my last comment was ambiguous. I'm exposing `FuntionInfo` and `ConstantInfo` on purpose, but they both should reside in the top level package and I forgot to move them there when I intentionally exposed them.

Also, I don't intend to replace the AST, but I might change the signature of AstBuilders methods or its return types.

**Comment by FabianNitsche** [comment]
Maybe it is generally a good idea to go towards dependency injection and use the current Evaluator as a factory for easy composition? This could allow for more extensibility, like with issue #48 (Deeper Optimization), DI could allow to have different optimizers for different scenarios. Then one does not need `optimization = false` but rather injects a `NoOptimization` instance. The same for caching, look-ups etc.

BTW: I do not recommend to derive from IEnumerable, it restricts extensibility. Imagine, you ever need to provide two collections to the user of the interface. It will be confusing to the user to have one collection as the instance itself an another as a property. Better to put collections to properties. Additionally, it can be unclear what the IEnumerable will provide the user, since one may just have the name of the instance, so context may go missing.

**Comment by vlow** [comment]
> Maybe it is generally a good idea to go towards dependency injection and use the current Evaluator as a factory for easy composition? This could allow for more extensibility, like with issue https://github.com/adletec/sonic/issues/48 (Deeper Optimization), DI could allow to have different optimizers for different scenarios. Then one does not need optimization = false but rather injects a NoOptimization instance. The same for caching, look-ups etc.

That's worth exploring, but I feel like we're starting to get lost in thoughts about the general architecture and further away from the initial issue. Let's try to focus on a quick solution for your scalar substitution and put the architecture discussion in its own separate issue.

Would the solution proposed above work for you? I am pretty confident that this solution would be possible without burdening us with debts for a general adjustment in the API design.

> BTW: I do not recommend to derive from IEnumerable, it restricts extensibility. Imagine, you ever need to provide two collections to the user of the interface. It will be confusing to the user to have one collection as the instance itself an another as a property. Better to put collections to properties. Additionally, it can be unclear what the IEnumerable will provide the user, since one may just have the name of the instance, so context may go missing.

Completely agree, this isn't a _sonic_ design decision but technical debt from the jace code base. The decision to expose `IEnumerable<>` instead of `IFunctionRegistry` was a somewhat dirty shortcut to prevent making this crude construct part of the public API. It might be a good opportunity to get rid of this idiosyncrasy.

**Comment by FabianNitsche** [comment]
Sorry, yes I got a little carried away.
Sometimes it is better to sleep over a problem. Generally, Evaluator is rather a Factory, all its components can be generated individually, so I may just wrap the parts in my own class. And solve my problems myself... :D 
The `IConstantLookupProvider` does not help me much I can recreate the Repositories from the exposed `IFunctionInfo`s and `IConstantInfo`s. 

Thank you for your time.
